Improvements on BTAction

ajtudela opened the pull request:
I made some improvements to the BTActions:
- I added ```server_timeout``` as a parameter.
- I split timeouts between server and wait, and I added a new ```wait_for_service_timeout``` parameter.
- I removed the rate from the ActionExecutorClient constructor to simplify its usage, it was only used in bt_actions, and is now a fully set parameter. I renamed the internal member to 'period_' for clarification. The default is 200ms / 5.0Hz.
- I added a new utils to function (from Davide in Nav2) to handle with the setting-getting of the ports / blackboard.
- The new parameters are set in the blackboard of the BT. Adding to the blackboard it makes all Nav2 nodes compatible with PlanSys2.

ajtudela committed:
Added server_timeout as parameter and blackboard

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Added wait_for_service_timeout param and blackboard

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Remove rate of ActionExecutorClient constructor and set as default parameter

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Added bt_loop_duration to the blackboard

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Fix test

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Fix more test

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Added explicit to constructor

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Fix explicit in test

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Minor fixes

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Fix some bt _action tests

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Added spin to test

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

fmrico commented on plansys2_bt_actions/src/plansys2_bt_actions/BTAction.cpp:
```diff
@@ -54,21 +54,42 @@ BTAction::on_configure(const rclcpp_lifecycle::State & previous_state)
+
+  // Put items in the blackboard
+  blackboard_->set<rclcpp::Node::SharedPtr>("node", client_node_);  // NOLINT
+  blackboard_->set<std::chrono::milliseconds>("server_timeout", default_server_timeout_);  // NOLINT
```
Why `// NOLINT`? I prefer to fix sthe tyle

fmrico commented on plansys2_bt_actions/include/plansys2_bt_actions/BTAction.hpp:
```diff
@@ -34,9 +34,7 @@ namespace plansys2
 class BTAction : public plansys2::ActionExecutorClient
 {
 public:
-  explicit BTAction(
```
I prefer to maintain both constructors for backward compatibility. You can add a warning of deprecation in the old constructor.

fmrico commented on plansys2_bt_actions/test/behavior_tree/Move.cpp:
```diff
@@ -32,7 +32,7 @@ Move::Move(
 : plansys2::BtActionNode<test_msgs::action::Fibonacci>(xml_tag_name, action_name, conf)
 {
-  rclcpp_lifecycle::LifecycleNode::SharedPtr node;
+  rclcpp::Node::SharedPtr node;
```
This is very sensible. Currently, these BT nodes are used in action performers, which are cascade_lifecycle...

ajtudela replied:
I understand the need for the main nodes to be lifecycle nodes. However, the node within the BTActionNode does not have to be a lifecycle node because the action is executed when the ActionExecutorClient instructs it to. In fact, the advantages of lifecycle nodes (configure, activate, etc.) are never utilized in the BTActionNode.

A node client inside the BTActionNode works the same way: 
https://github.com/PlanSys2/ros2_planning_system/compare/rolling...grupo-avispa:ros2_planning_system:improve_bt_actions#diff-d7c4293a0e6ee3223f93bd6fcd115bb5fe9b52f2049ac5e266010e9a7c47fba8R78-R79

fmrico commented on plansys2_executor/test/unit/executor_test.cpp:
```diff
@@ -224,13 +224,16 @@ TEST(executor, action_executor_client)
   auto test_node = rclcpp_lifecycle::LifecycleNode::make_shared("test_node");
   auto aux_node = rclcpp_lifecycle::LifecycleNode::make_shared("aux_node");
 
```
As I said in the previous comment, let's maintain both constructors, so duplicate the test and modify the new test, and maintain this.

ajtudela committed:
Remove nolint

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Added deprecated constructor

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Revert lifecycle

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela commented:
I added a ```deprecated```attribute to the old constructor and a new test that use it. Other tests are updated with the new constructor.
I reverted the rclcppp_lifecycle inside the BT.

ajtudela committed:
Fix print

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Added JSON utils

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Added a BT for services

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

ajtudela committed:
Update BTServiceNode.hpp

ajtudela committed:
Fix BTService and BT-JSON

Signed-off-by: Alberto Tudela <ajtudela@gmail.com>

fmrico commented:
LGTM

Merging!!! :rocket:
